modules/affinity/src/main/java/org/apache/ignite/internal/affinity/Assignments.java

```diff
@@ -0,0 +1,228 @@
+     */
+    public void set(int part, Set<Assignment> assignment) {
+        assignments.set(part, assignment);
+    }
+
+    /**
+     * Bytes representation of the assignments.
+     *
+     * @return Bytes representation of the assignments.
+     */
+    public byte[] bytes() {
+        Map<String, Integer> consistentIds = new LinkedHashMap<>();
+
+        int idx = 0;
+        for (Set<Assignment> assignmentSet : assignments) {
+            for (Assignment a : assignmentSet) {
+                if (consistentIds.putIfAbsent(a.consistentId(), idx) == null) {
+                    idx++;
+                }
+            }
+        }
+
+        byte[] consistentIdsBytes = collectionToBytes(consistentIds.keySet(), String::getBytes);
+
+        byte[] assignmentsBytes = collectionToBytes(
+                assignments,
+                set -> collectionToBytes(set, a -> assignmentToBytes(a, consistentIds::get))
+        );
+
+        ByteBuffer buf = ByteBuffer.allocate(Integer.BYTES * 2 + consistentIdsBytes.length + assignmentsBytes.length);
+
+        buf.putInt(consistentIdsBytes.length);
+        buf.put(consistentIdsBytes);
+        buf.putInt(assignmentsBytes.length);
+        buf.put(assignmentsBytes);
+
+        return buf.array();
+    }
+
+    /**
+     * Creates {@link Assignments} from the given byte array.
+     *
+     * @param bytes Byte array representation of the assignments.
+     * @return Assignments.
+     */
+    public static Assignments fromBytes(byte[] bytes) {
+        ByteBuffer buf = ByteBuffer.wrap(bytes);
+
+        int consistentIdsBytesLength = buf.getInt();
+        byte[] consistentIdsBytes = new byte[consistentIdsBytesLength];
+        buf.get(consistentIdsBytes);
+
+        int assignmentsBytesLength = buf.getInt();
+        byte[] assignmentsBytes = new byte[assignmentsBytesLength];
+        buf.get(assignmentsBytes);
+
+        List<String> consistentIds = bytesToList(consistentIdsBytes, b -> new String(b, StandardCharsets.UTF_8));
+
+        List<Set<Assignment>> assignments = bytesToList(
+                assignmentsBytes,
+                b -> new HashSet<>(bytesToList(b, ab -> bytesToAssignment(ab, consistentIds::get)))
+        );
+
+        return new Assignments(assignments);
+    }
+
+    /**
+     * Serializes collection to bytes.
+     *
+     * @param collection Collection.
+     * @param transform Tranform function for the collection element.
+     * @return Byte array.
+     */
+    private <T> byte[] collectionToBytes(Collection<T> collection, Function<T, byte[]> transform) {
+        int bytesObjects = 0;
+        List<byte[]> objects = new ArrayList<>();
+
+        for (T o : collection) {
+            byte[] b = transform.apply(o);
+            objects.add(b);
+            bytesObjects += b.length;
+        }
+
+        bytesObjects += Short.BYTES * (objects.size() + 1);
+
+        ByteBuffer buf = ByteBuffer.allocate(bytesObjects);
+
+        buf.putShort((short) objects.size());
```

Comment:
Like here, there must be a check that the conversion is safe, otherwise the structure is broken

Reply:
For example, you're assuming that there can't be more than 65k partitions, which will not be true in the future. So we should use int for the header instead of short

Reply:
agree